### ISSUE TITLE
Difficulties with implementing evolutionary search algorithm in Haiku

### ISSUE THREAD

**Issue by ttt733** [deepmind/dm-haiku#196]
Hi there. In an effort to get familiar with Haiku (and Jax,) I thought that I would try to modify one of [the examples](https://github.com/deepmind/dm-haiku/tree/main/examples/transformer) by changing it from gradient descent to using a simple evolutionary search. However, there were a couple issues that I was not sure how to figure out after reading the documentation. If anyone has an idea of what I'm missing, I'd appreciate any pointers.

1. Slow module initialization. My initial idea was to implement a custom initializer that took the weight arrays from two parent modules' params and did a crossover operation on them. I planned to just use that as a drop-in replacement for the initializers used in the example. The problem was that every "reproduction" phase slowed training to a crawl, during the initialization of the new modules. I changed from CheckpointingUpdater to Updater to avoid the blocking done in that example, but it didn't seem to have an effect. Is there anything specific that needs to be done in order to initialize new modules quickly? (I'll attach some of the code I changed in the main loop down below in case that makes it more obvious where my modifications are going wrong, or what else I would need to add.)
2. Manual parameter updates. As a workaround, I thought I might try simply updating the weights of the parent modules directly rather than initializing new children, and just keep the population size fixed. However, I couldn't find any examples of how to do this. Would I just want to call `hk.set_state('params', {...})`?

Code that might be relevant to my first question:
```
# In forward_fn
transformer = model_evolve.EvolvedTransformer(
        num_heads=num_heads, num_layers=num_layers, dropout_rate=dropout_rate,
        parent_1_params=member['parent_1_params'], parent_2_params=member['parent_2_params'])
# ...
def add_training_functions(vocab_size, member, rng, data):
    # Set up the model, loss, and updater.
    forward_fn = build_forward_fn(vocab_size, FLAGS.d_model, FLAGS.num_heads,
                                    FLAGS.num_layers, FLAGS.dropout_rate, member)
    forward_fn = hk.transform(forward_fn)
    loss_fn = functools.partial(lm_loss_fn, forward_fn.apply, vocab_size)
    updater = Updater(forward_fn.init, loss_fn)
    state = updater.init(rng, data)
    member['forward_fn'] = forward_fn
    member['loss_fn'] = loss_fn
    member['updater'] = updater
    member['state'] = state
    return member

def main(_):
  # Create the dataset.
  train_dataset, vocab_size = dataset.load(FLAGS.batch_size,
                                           FLAGS.sequence_length)

  logging.info('Initializing parameters...')
  rng = jax.random.PRNGKey(428)
  data = next(train_dataset)

  # Initialize population members.
  members = []
  for _ in range(POPULATION_SIZE):
    members.append({
        'parent_1_params': None,
        'parent_2_params': None,
        'params': None,
        'total_loss': 0,
    })
  for member_idx in range(len(members)):
    # Set up the model, loss, and updater.
    members[member_idx] = add_training_functions(vocab_size, members[member_idx], rng, data)

  logging.info('Starting train loop...')
  prev_time = time.time()
  for step in range(MAX_STEPS):
    if step != 0 and step % STEPS_PER_GENERATION == 0:
        # Reproduce before continuing
        # First, find the top performer
        best_loss = np.Inf
        best_member = None
        for member in members:
            member_loss = member['metrics']['loss']
            if member_loss < best_loss:
                best_member = member
                best_loss = member_loss
        for member_idx in range(len(members)):
            # Generate a child with the top performer
            member = members[member_idx]
            new_member = {
                'parent_1_params': member['params'],
                'parent_2_params': best_member['params'],
                'params': None,
            }
            members[member_idx] = add_training_functions(vocab_size, new_member, rng, data)
    data = next(train_dataset)
    for member_idx in range(len(members)):
        member = members[member_idx]
        state, metrics = member['updater'].update(member['state'], data)
        member['state'] = state
        member['metrics'] = metrics
        # We use JAX runahead to mask data preprocessing and JAX dispatch overheads.
        # Using values from state/metrics too often will block the runahead and can
        # cause these overheads to become more prominent.
        if step % LOG_EVERY == 0:
            steps_per_sec = LOG_EVERY / (time.time() - prev_time)
            prev_time = time.time()
            metrics.update({'steps_per_sec': steps_per_sec})
            metrics.update({'member_number': member_idx})
            logging.info({k: float(v) for k, v in metrics.items()})
```
And just in case, here's my attempt at a crossover initializer, called from within the transformer:
```
class CrossoverInitializer(hk.initializers.Initializer):
  """Initializes with a crossover of two sets of parent parameters."""

  def __init__(
    self,
    parent_1: ndarray,
    parent_2: ndarray,
    mutation_scale=.003,
  ):
    assert parent_1.shape == parent_2.shape
    self.p_1 = parent_1
    self.p_2 = parent_2
    self.mutation_scale = mutation_scale
    

  def __call__(self, shape: Sequence[int], dtype: Any) -> jnp.ndarray:
    parent_1 = self.p_1
    parent_2 = self.p_2
    crossover_idx = int(jax.random.uniform(hk.next_rng_key(), ()) * parent_1.shape[0])
    child = jnp.concatenate((parent_1[:crossover_idx], parent_2[crossover_idx:]))
    mutation = jax.random.normal(hk.next_rng_key(), shape, dtype) * self.mutation_scale
    return child + mutation
```

**Comment by tomhennigan** [comment]
Hi @ttt733 , from looking at your code quickly (I have not found time to run it) you are creating `POPULATION_SIZE` different `Updater` instances.

Each of these has an `init` and `update` function which will get `jit` compiled by JAX. I suspect this is why training is taking so long (even if JIT compiling each function takes a few seconds, if your population is sufficiently large this will quickly be minutes or hours of compilation). You may find this guide useful https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html

You probably want to refactor the code such that you have a single init function that creates a member, and a single update function that updates a member. You should be able to jit these just once (and pass the member in as an argument to the function) which will mean your code will run fast and you will spend little time in compilation. You may find that the `Updater` pattern in the transformer example is not amenable to this, but there are many ways to write Haiku/JAX code, for example our imagenet example uses a top level function for `initial_state` and `train_step`, it may serve as a useful reference. https://github.com/deepmind/dm-haiku/blob/main/examples/imagenet/train.py#L193

**Comment by ttt733** [comment]
That's almost definitely it. I was looking for code that wasn't going through JIT, but code going through it repeatedly is more likely in this case. Thanks for taking the time to look - I'll give some of the other example patterns a try.

**Comment by ttt733** [comment]
In case anyone winds up here later, here's the code I landed on for the training and initialization:
https://gist.github.com/ttt733/d574ea83f89181043c03c95e5c680403

I've not included the models because (no surprise) crossover in this case wasn't actually very effective, but hopefully it shows you one way to initialize networks faster. On my single GPU machine, I was executing SGD on the example module at about 60 steps per second, and I was able to run this loop on my customized module at about 40-ish steps per second including reproduction (6 seconds per generation with a population size of 24.) You could probably optimize it even more than that, and obviously using pmap here would be better if you had multiple training devices.
